rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java

```diff
@@ -492,35 +491,11 @@ private int retryGetEndStatus(URI endPath, boolean compensate) {
                                     compensate, target.getUri(), status);
 
                             if (forgetURI != null) {
-                                try {
-                                    // let the participant know he can clean up
-                                    WebTarget target2 = client.target(forgetURI);
-                                    AsyncInvoker asyncInvoker2 = target2.request()
-                                            .header(LRA_HTTP_CONTEXT_HEADER, lraId.toASCIIString())
-                                            .property(LRA_HTTP_PARENT_CONTEXT_HEADER, parentId) // make the context available to participants
-                                            .header(LRA_HTTP_RECOVERY_HEADER, recoveryURI.toASCIIString())
-                                            .property(LRA_HTTP_CONTEXT_HEADER, lraId)  // make the context available to the jaxrs filters
-                                            .async();
-
-                                    Future<Response> asyncResponse2 = getAsyncResponse(
-                                            target2, DELETE.class.getName(), asyncInvoker2, compensatorData);
-
-                                    if (asyncResponse2.get(PARTICIPANT_TIMEOUT, TimeUnit.SECONDS).getStatus() ==
-                                            Response.Status.OK.getStatusCode()) {
-                                        return TwoPhaseOutcome.FINISH_OK;
-                                    }
-                                } catch (Exception e) {
-                                    if (LRALogger.logger.isInfoEnabled()) {
-                                        LRALogger.logger.infof("LRARecord.doEnd forget URI %s is invalid (%s)",
-                                                forgetURI, e.getMessage());
-                                    }
-
-                                    // TODO write a test to ensure that recovery only retries the forget request
-                                    return TwoPhaseOutcome.HEURISTIC_HAZARD; // force recovery to keep retrying
-                                } finally {
-                                    Current.pop();
+                                if (invokeForget(client, forgetURI, lraId, parentId, recoveryURI)) {
+                                    return TwoPhaseOutcome.FINISH_OK;
+                                } else {
+                                    return TwoPhaseOutcome.HEURISTIC_HAZARD;
```

Comment:
Are there any other locations where failure to invokeForget should cause a HEURISTIC_HAZARD. For instance over [here](https://github.com/mmusgrov/narayana/blob/d88fe2fcef356a333946a02d282f849e0f2094f4/rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java#L271) it seems it was proposed to return the HEURISTIC_HAZARD but in the current patch forget is not called at all (at least from the [similar](https://github.com/jbosstm/narayana/blob/9380ad2ffdfb0716aae6ab743a4b6dca81a3d97d/rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java#L261) place, perhaps it is called elsewhere by the refactor but then I am not sure how the refactor reports the H_H?)

Reply:
there is no need because already completed nested participants with forget methods are tracked in the same manner as afterLRA listeners - only in Transaction. So LRARecord will no longer be invoked. 

This leads me again to the issue that I realized because of your comments and I think that even afterLRA listeners may be doing something unexpected in the moment. I will report this in a new JBTM (as this PR is not the right place to track it) when I'll have the test ready.